hail/src/main/scala/is/hail/expr/ir/Exists.scala

```diff
@@ -31,16 +31,54 @@ object Exists {
 }
 
 object Forall {
-  def apply(node: BaseIR, visitor: BaseIR => Boolean): Boolean =
-    !Exists(node, n => !visitor(n))
+  def apply(node: IR, visitor: IR => Boolean): Boolean =
+    !Exists.inIR(node, n => !visitor(n))
+}
+
+object IsScanResult {
+  def apply(root: IR): Boolean = root match {
+    case _: ApplyScanOp => true
+    case AggFilter(_, _, isScan) => isScan
+    case AggExplode(_, _, _, isScan) => isScan
+    case AggGroupBy(_, _, isScan) => isScan
+    case AggArrayPerElement(_, _, _, _, isScan) => isScan
+    case _ => false
+  }
+}
+
+object IsAggResult {
+  def apply(root: IR): Boolean = root match {
+    case _: ApplyAggOp => true
+    case AggFilter(_, _, isScan) => !isScan
+    case AggExplode(_, _, _, isScan) => !isScan
+    case AggGroupBy(_, _, isScan) => !isScan
+    case AggArrayPerElement(_, _, _, _, isScan) => !isScan
+    case _ => false
+  }
 }
 
 object ContainsAgg {
-  def apply(root: BaseIR): Boolean = Exists(root, _.isInstanceOf[ApplyAggOp])
+  def apply(root: IR): Boolean = IsAggResult(root) || (root match {
```

Comment:
The previous implementations were wrong in that they would recur into the value children of things like TableAggregate, even if that wasn't part of the same scope.